### ISSUE TITLE
[RFC] Low-level speed optimizations for PowerSGD

### ISSUE THREAD

**Issue by tvogels** [pytorch/pytorch#65813]
### Communication compression

PyTorch features several [DDP Communication Hooks](https://pytorch.org/docs/stable/ddp_comm_hooks.html) that compress messages exchanged between workers in distributed optimization. If communication time is a bottleneck, these hooks can speed up distributed training. Of course, compression is only beneficial if the *time required to compress the messages* is significantly smaller than the *time spent communicating*. Speeding up compression times can open up communication savings to a wider range of models and hardware.

### PowerSGD

In training problems with a strong communication bottleneck, the current [PowerSGD](https://arxiv.org/abs/1905.13727) hook in PyTorch already improves training times (@SciPioneer), but a [recent paper](https://arxiv.org/pdf/2103.00543.pdf) argues that in many settings, the gains in communication do not yet weigh up to the added compression time. 

### Proposed optimizations 
The recent [DALL-E paper](https://arxiv.org/abs/2102.12092) uses PowerSGD for large-scale distributed training, and the paper's appendix contains many recommendations on how to implement the algorithm efficiently. The most actionable recommendation is 
- the creation of a specialized CUDA kernel for the orthogonalization of matrices with many rows and few columns.  

Orthogonalization is the most expensive step in PowerSGD compression and based on timing results from the DALL-E authors, there is a potential for speedups up to 100x in this operation. 

### Benefits
- With faster compression times, compression will generate speedups for more models and with faster communication hardware.
- A faster orthogonalization operation will allow PowerSGD to use higher 'rank's (more accurate compression). Currently, such large ranks would make compression too slow. With more accurate compression, we can avoid drops in model accuracy.

**Comment by SciPioneer** [comment]
Per discussion offline, @tvogels  is going to do some fast prototyping and provide some early results first.

CC: @pritamdamania @cbalioglu

**Comment by cbalioglu** [comment]
Thanks for the proposal @tvogels! Sounds interesting. A while ago internally we raised the same concern about Gram-Schmidt's sequential nature. Looking forward to your prototype. In case we see some promising results, we can definitely consider integrating it to PowerSGD.

**Comment by tvogels** [comment]
@cbalioglu great, thanks! My CUDA is pretty rusty, so if anybody with GPU expertise wants to have a chat, let me know.

**Comment by tvogels** [comment]
@cbalioglu , yes I discussed this with Scott Gray, one of the authors of DALL-E. They use Householder orthogonalization based on [this video](https://www.youtube.com/watch?v=d-yPM-bxREs).

**Comment by cbalioglu** [comment]
@tvogels great! Let me get back to you on Monday.

**Comment by tvogels** [comment]
@cbalioglu if you want, maybe we can have a quick chat about this later this week? You can reach me by email with `{first}.{last} at {epfl}.ch`.

**Comment by tvogels** [comment]
The proposal #74907 should also be considered in this effort.
